New parser

melaniecebula opened the pull request:
@mafintosh: thoughts?

This is a possible way to address https://github.com/mafintosh/hackfile/issues/7

Commit:
Nest all one-line commands by default to handle ambiguity

mafintosh commented on test/index.js:
```diff
@@ -19,19 +19,21 @@ tape('basic nesting', function(t) {
 tape('advanced', function(t) {
-  var result = [["pipeline",["foo", "run echo hi",["map",["curl http","a","b","c",["pipe",["echo hello","transform","cat"]]]],["run",["echo","hi"]]]],["pipeline",["bar","run echo hi"]],["run",["echo bye"]]]
+  var result = [ ['pipeline',['foo',['run',['echo hi']],['map',[['curl',['http']],'a','b','c',['pipe',[['echo',['hello']],'transform','cat']]]],['run',['echo','hi']]]],
+    ['pipeline',['bar',['run',['echo hi']]]],
```
shouldn't this be `[..., ['run', ['echo', ['hi']]]]`?

mafintosh replied:
or is it because this was parsed from 

```
run echo hi
```

and not

```
run
  echo hi
```

?

melaniecebula replied:
yes, I have been treating one-line commands differently.

For example:
```
run echo hi
  echo hi
  echo hi

run
  echo hi
  echo hi
  echo hi

run echo hi
```

output (stringified):
```
[ '["run",[["echo",["hi"]],["echo",["hi"]],["echo",["hi"]]]]',
  '["run",[["echo",["hi"]],["echo",["hi"]],["echo",["hi"]]]]',
  '["run",["echo hi"]]' ]
```

But I can easily change it so that the output is:
```
[ '["run",[["echo",["hi"]],["echo",["hi"]],["echo",["hi"]]]]',
  '["run",[["echo",["hi"]],["echo",["hi"]],["echo",["hi"]]]]',
  '["run",["echo",  ["hi"]]' ]
```
(so the one-line command is treated more like the other nested commands now)

I'm a fan of this.  What do you think?

mafintosh replied:
I like it as well. We should do that.

mafintosh commented on test/index.js:
```diff
@@ -19,19 +19,21 @@ tape('basic nesting', function(t) {
-  var result = [["pipeline",["foo", "run echo hi",["map",["curl http","a","b","c",["pipe",["echo hello","transform","cat"]]]],["run",["echo","hi"]]]],["pipeline",["bar","run echo hi"]],["run",["echo bye"]]]
+  var result = [ ['pipeline',['foo',['run',['echo hi']],['map',[['curl',['http']],'a','b','c',['pipe',[['echo',['hello']],'transform','cat']]]],['run',['echo','hi']]]],
+    ['pipeline',['bar',['run',['echo hi']]]],
+    ['run',['echo bye']] ]
```
same as above

melaniecebula commented:
changed so one-line commands are treated more like other nested commands, added another test, changed README to update example, minor changes (use shift instead of splice)

mafintosh commented:
I fixed the nested example in the README. Released in 4.0.0
